src/event/signal_child_handler.cr

```diff
@@ -0,0 +1,60 @@
+# :nodoc:
+# Singleton that handles SIG_CHLD and queues events for Process#waitpid.
+# Process.waitpid uses this class for nonblocking operation.
+class Event::SignalChildHandler
+  def self.instance
+    @@instance ||= new
+  end
+
+  alias ChanType = BufferedChannel(Process::Status?)
+
+  def initialize
+    @pending = Hash(LibC::PidT, Process::Status).new
+    @waiting = Hash(LibC::PidT, ChanType).new
+  end
+
+  def after_fork
+    @pending.clear
+    @waiting.each { |pid, chan| chan.send(nil) }
+    @waiting.clear
+  end
+
+  def trigger
+    loop do
+      pid = LibC.waitpid(-1, out exit_code, LibC::WNOHANG)
```

Comment:
I speculate that passing -1 is going to interact badly with threads.  The exit notification may end up on the wrong thread.